File: apps/youtube/extensions/speed/src/main/java/app/reseam/youtube/speed/RememberPlaybackSpeed.java
// SPDX-FileCopyrightText: 2026 AunAli K. <hello@auna.li>
// SPDX-FileCopyrightText: 2026 ReVanced contributors
// SPDX-License-Identifier: GPL-3.0-or-later

package app.reseam.youtube.speed;

import app.reseam.runtime.settings.ReseamSettings;
import app.reseam.youtube.core.Logger;
import app.reseam.youtube.core.Settings;
import app.reseam.youtube.video.VideoInformation;

/** Applies the default speed once per video and persists explicit selections. */
public final class RememberPlaybackSpeed {
    private static volatile boolean newVideoStarted;

    private RememberPlaybackSpeed() {}

    public static void newVideoStarted(VideoInformation.PlaybackController controller) {
        newVideoStarted = true;
    }

    /** Called when YouTube initializes its speed values: after a video loads, and after every selection. */
    public static void applyDefaultPlaybackSpeed() {
        if (!newVideoStarted) return;
        newVideoStarted = false;
        final float speed = defaultSpeed();
        if (speed <= 0.0f) return;
        VideoInformation.overridePlaybackSpeed(speed);
        Logger.debug(() -> "Default playback speed applied: " + speed);
    }

    /** The configured default speed, or 0 when none is set. */
    public static float defaultSpeed() {
        return parse(Settings.getString("playback_speed_default", "0"));
    }

    public static void userSelectedPlaybackSpeed(float speed) {
        if (!Settings.getBoolean("remember_playback_speed", false)) return;
        ReseamSettings.setString("you_tube_settings.playback_speed_default", Float.toString(speed));
        Logger.debug(() -> "Remembered playback speed: " + speed);
    }

    private static float parse(String value) {
        try {
            float speed = Float.parseFloat(value);
            return speed > 0 && speed <= CustomPlaybackSpeed.PLAYBACK_SPEED_MAXIMUM ? speed : 0.0f;
        } catch (Exception exception) {
            return 0.0f;
        }
    }
}

File: apps/youtube/patch/src/main/kotlin/app/reseam/patches/youtube/video/speed/RememberPlaybackSpeed.kt
// SPDX-FileCopyrightText: 2026 AunAli K. <hello@auna.li>
// SPDX-FileCopyrightText: 2026 ReVanced contributors
// SPDX-License-Identifier: GPL-3.0-or-later

package app.reseam.patches.youtube.video.speed

import app.reseam.patch.ExtClass
import app.reseam.patch.Type
import app.reseam.patch.before
import app.reseam.patch.patch
import app.reseam.patch.settings.section
import app.reseam.patches.youtube.core.YOUTUBE
import app.reseam.patches.youtube.core.YouTubeSettings
import app.reseam.patches.youtube.core.YouTubeSettingsPages
import app.reseam.patches.youtube.core.youTubeSettings
import app.reseam.patches.youtube.internal.onCreateHook
import app.reseam.patches.youtube.internal.userSelectedPlaybackSpeedHook
import app.reseam.patches.youtube.internal.videoInformationHook

object RememberPlaybackSpeed : ExtClass("app.reseam.youtube.speed.RememberPlaybackSpeed") {
    val newVideoStarted = static("newVideoStarted", "app.reseam.youtube.video.VideoInformation\$PlaybackController")
    val applyDefaultPlaybackSpeed = static("applyDefaultPlaybackSpeed")
    val userSelectedPlaybackSpeed = static("userSelectedPlaybackSpeed", Type.Float)
}

val rememberPlaybackSpeed = patch("Remember playback speed") {
    description("Applies a default playback speed and remembers explicit selections.")
    compatibleWith(YOUTUBE)
    dependsOn(youTubeSettings, videoInformationHook, customPlaybackSpeed)
    settings(youTubeSettings, section(YouTubeSettingsPages.Video, "Playback speed", YouTubeSettings.playbackSpeedDefault, YouTubeSettings.rememberPlaybackSpeed))

    execute {
        onCreateHook(RememberPlaybackSpeed.newVideoStarted)
        userSelectedPlaybackSpeedHook(RememberPlaybackSpeed.userSelectedPlaybackSpeed)
        oldPlaybackSpeedMenu.before { call(RememberPlaybackSpeed.applyDefaultPlaybackSpeed) }
    }
}
